nbviewer/app.py

```diff
@@ -159,106 +172,138 @@ def init_tornado_application(self):
-            localfile_any_user=options.localfile_any_user,
-            fetch_kwargs=fetch_kwargs,
-            mathjax_url=options.mathjax_url,
-            rate_limiter=rate_limiter,
-            statsd_host=options.statsd_host,
-            statsd_port=options.statsd_port,
-            statsd_prefix=options.statsd_prefix,
-            base_url=base_url,
-            google_analytics_id=os.getenv('GOOGLE_ANALYTICS_ID'),
-            hub_api_token=os.getenv('JUPYTERHUB_API_TOKEN'),
-            hub_api_url=os.getenv('JUPYTERHUB_API_URL'),
-            hub_base_url=os.getenv('JUPYTERHUB_BASE_URL'),
-            ipywidgets_base_url=options.ipywidgets_base_url,
-            jupyter_widgets_html_manager_version=options.jupyter_widgets_html_manager_version,
-            jupyter_js_widgets_version=options.jupyter_js_widgets_version,
-            content_security_policy=options.content_security_policy,
-            binder_base_url=options.binder_base_url,
+                  config=self.config,
+                  index=self.index,
+                  max_cache_uris=self.max_cache_uris,
+                  static_path=self.static_path,
+                  static_url_prefix=self.static_url_prefix,
         )
-    
+        # input computed properties to settings
+        settings.update(
+                  base_url=self.base_url,
+                  cache=self.cache,
+                  client=self.client,
+                  fetch_kwargs=self.fetch_kwargs,
+                  formats=self.formats,
+                  frontpage_setup=self.frontpage_setup,
+                  jinja2_env=self.env,
+                  pool=self.pool,
+                  rate_limiter=self.rate_limiter,
+        )
+        # input settings from CLI options
+        settings.update(
+                  binder_base_url=options.binder_base_url,
+                  cache_expiry_max=options.cache_expiry_max,
+                  cache_expiry_min=options.cache_expiry_min,
+                  content_security_policy=options.content_security_policy,
+                  default_format=options.default_format,
+                  ipywidgets_base_url=options.ipywidgets_base_url,
+                  jupyter_js_widgets_version=options.jupyter_js_widgets_version,
+                  jupyter_widgets_html_manager_version=options.jupyter_widgets_html_manager_version,
+                  localfile_any_user=options.localfile_any_user,
+                  localfile_follow_symlinks=options.localfile_follow_symlinks,
+                  localfile_path=os.path.abspath(options.localfiles),
+                  mathjax_url=options.mathjax_url,
+                  provider_rewrites=options.provider_rewrites,
+                  providers=options.providers,
+                  render_timeout=options.render_timeout,
+                  statsd_host=options.statsd_host,
+                  statsd_port=options.statsd_port,
+                  statsd_prefix=options.statsd_prefix,
```

Comment:
Is the goal ultimately to stop using tornado `options` for command line parsing and switch completely to using traitlet configuration on the command line (e.g., `--NBViewer.cache_expiry_max=12345`) and/or in a config file?

I'm asking to understand:

1. if nbviewer will always have a dependence on two configuration systems going forward: tornado.options and traitlets.
2. if the goal is to move everything to traitlets, then when should we cut a final 1.x pypi release of nbviewer fixes up to that point, because we're about to drastically change the command line "API".

I'm sure that seeing the rest of the PRs you have queued up will help answer the question but I figured I'd get these thoughts out as part of the review.

Reply:
The goal is to move everything to traitlets (see https://github.com/jupyter/nbviewer/pull/847 where this was finally done). 

I was actually able (I think) to move everything to traitlets without affecting the command line "API", so hopefully that won't be an issue. (It's possible using aliases and flags.)

This would make NBViewer a first-class Jupyter application like you mentioned, plus the configuration system is more extensible and flexible, since now one can use both the command line as well as the config file.

Reply:
So e.g. you can do `--cache_expiry_max=12345` _or_ `--NBViewer.cache_expiry_max=12345`. (As well as add a line `c.NBViewer.cache_expiry_max = 12345` to `nbviewer_config.py`, with the command line settings taking precedence over the config file ones. It's even possible to change the config file via the command line.)

I'm not sure what a "cut a release" means or what that would entail. Sorry for being dumb; I haven't ever been this involved in an OSS project before and am probably not familiar with all of the terminology.

Reply:
I forgot about the alias support in traitlets. I believe you're correct that we can switch to traitlets without intentionally changing the CLI parameters. There might be some gotchas in how things are parsed, but we can deal with those as we find them.

> I'm not sure what "cut a release" means or what that would entail. 

By "cut a release" I meant making a tag here on GitHub and pushing a package to PyPI before we make any substation changes to the CLI. However, given that you're using aliases, that might not be necessary to do up front.

> Sorry for being dumb; I haven't ever been this involved in an OSS project before and am probably not familiar with all of the terminology.

You're not being dumb and you're doing great. Thank you for all the work!